homework13/src/main/java/ua/lviv/home/daos/BucketDao.java

```diff
@@ -0,0 +1,117 @@
+package ua.lviv.home.daos;
+
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.log4j.Logger;
+import ua.lviv.home.EntityManagerUtils;
+import ua.lviv.home.enteties.Bucket;
+
+import javax.persistence.EntityManager;
+import java.util.ArrayList;
+import java.util.List;
+
+public class BucketDao implements CRUD<Bucket> {
+
+    private static final Logger LOG = Logger.getLogger(BucketDao.class);
+
+    private static String READ_ALL = "select b from Bucket b";
+    private static String READ_ALL_BY_USER_ID = "select b from Bucket b where b.userId in (:userId)";
+    private static String READ_BY_ID = "select b from Bucket b where b.id =:id";
+
+    public BucketDao() {
+    }
+
+    @Override
+    public Bucket insert(Bucket bucket) {
+        String message = String.format("Will create a bucket for userId=%d and productId=%d",
+                bucket.getUserId(), bucket.getProductId());
+        LOG.debug(message);
+
+        try {
+            EntityManager entityManager = EntityManagerUtils.getEntityManager();
+            entityManager.getTransaction().begin();
+            entityManager.persist(bucket);
+            entityManager.flush();
+            entityManager.getTransaction().commit();
+
+        } catch (Exception e) {
+            String errorMessage = String.format("Fail to create a bucket for userId=%d and productId=%d",
+                    bucket.getUserId(), bucket.getProductId());
+            LOG.error(errorMessage, e);
+        }
+        return bucket;
+    }
+
+    @Override
+    public Bucket read(int id) {
+        Bucket bucket = null;
+        try {
+            EntityManager entityManager = EntityManagerUtils.getEntityManager();
+            entityManager.getTransaction().begin();
```

Comment:
Actually, reading from DB could be without transaction, so opening one is redundant here.

Reply:
done)

Reply:
If I take away the transaction, the bucket in the database and the site is not updated